Update object handlers

iluuu1994 opened the pull request:
This file only exists for the PHP 5 part of the book, I wonder if it's time to merge those and just make `master` match the current `master` version of php-src itself.

nikic commented:
Updating the PHP 5 part with PHP 7 docs doesn't make sense. The object section would likely need a from scratch rewrite for PHP 7, because things changed a lot in that area.

iluuu1994 commented:
@nikic I only noticed that this was in the PHP 5 section after. Should I copy the whole chapter to the 7 section and remove anything that's no longer accurate? Does it make sense to keep the two versions around?

iluuu1994 commented:
@nikic I copied the handler descriptions to the PHP 7/8 section and removed anything else that might be outdated. There were also some handlers missing which I added. Let me know if this looks good to you. Could you grant me edit access to this repository? I won't merge anything without creating an PR first except for minor non-technical things like typos.

nikic commented on Book/php7/classes_objects/object_handlers.rst:
```diff
@@ -0,0 +1,137 @@
+to userland PHP. For example internal classes can have custom comparison and cast behavior.
+
+As the number of different object handlers is rather large we can only discuss examples (using the typed array
+implementation from the last section) for a few of them. For all the others only a short description is provided.
```
This paragraph (referencing typed array implemented) can be dropped.

nikic commented on Book/php7/classes_objects/object_handlers.rst:
```diff
@@ -0,0 +1,137 @@
+    These handlers correspond to the ``__get``, ``__set``, ``__isset`` and ``__unset`` methods. ``get_property_ptr_ptr``
+    is the internal equivalent of ``__get`` returning by reference. ``cache_slot`` is used to store the property
+    offset and ``zend_property_info``. ``zval *rv`` in ``read_property`` provides a place for temporary zvals that
+    are not stored in the object itself, like results of calls to ``__get``.
```
I think adding something like this would be helpful here, because `rv` ownership is the main tricky part:

> `read_property` may directly return a zval owned by the object, in which case its reference count should not be modified by `read_property`, and the caller should not release it. Alternatively, it may return `rv`, in which case the refcount should be incremented, and the caller is responsible for releasing the value.

nikic commented on Book/php7/classes_objects/object_handlers.rst:
```diff
@@ -0,0 +1,137 @@
+    zend_function *get_method(zend_object **obj_ptr, zend_string *method_name, const zval *key)
+
+    The ``get_method`` handler fetches the ``zend_function`` used to call a certain method. Either the ``method_name``
+    or ``key`` must be passed. ``key`` is assumed to be lower case.
```
It's not either/or. method_name must always be passed, key is optional.

nikic commented on Book/php7/classes_objects/object_handlers.rst:
```diff
@@ -0,0 +1,137 @@
+    This is just the internal way of implementing the ``Countable::count`` method. The function returns a
+    ``zend_result`` and assigns the value to the ``zend_long *count`` pointer.
+
+.. FIXME: Change return type of count_elements to zend_result to make it more obvious the count is not returned?
```
Makes sense

nikic commented on Book/php7/classes_objects/object_handlers.rst:
```diff
@@ -0,0 +1,137 @@
+.. c:member::
+    zend_string *get_class_name(const zend_object *zobj)
+
+    This handler is used to get the class name from an object. There should be little reason to overwrite it. The only
```
...for debugging contexts

nikic commented on Book/php7/classes_objects/object_handlers.rst:
```diff
@@ -0,0 +1,137 @@
+
+    This handler is used to get the class name from an object. There should be little reason to overwrite it. The only
+    occasion that I can think of where this would be necessary is if you choose to create a custom object structure that
+    does *not* contain the standard ``zend_object`` as a substructure. (This is entirely possible, but not usually done.)
```
This reference of non-standard zend_object is outdated, can be dropped.

nikic commented on Book/php7/classes_objects/object_handlers.rst:
```diff
@@ -0,0 +1,137 @@
+
+.. c:member::
+    zend_object *clone_obj(zend_object *old_object)
+    HashTable *get_gc(zend_object *zobj, zval **table, int *n)
```
Separate this out? They're not really related.

nikic commented on Book/php7/classes_objects/object_handlers.rst:
```diff
@@ -0,0 +1,137 @@
+.. FIXME: Change return type of count_elements to zend_result to make it more obvious the count is not returned?
+
+.. c:member::
+    int compare(zval *o1, zval *o2)
```
compare is mentioned both here and below.
